pkg/api/secretbootstrap/secretboostrap.go

```diff
@@ -91,6 +99,67 @@ func (c *Config) UnmarshalJSON(d []byte) error {
 	return c.resolve()
 }
 
+func (c *Config) MarshalJSON() ([]byte, error) {
+	target := &configWithoutUnmarshaler{
+		VaultDPTPPrefix:           c.VaultDPTPPrefix,
+		ClusterGroups:             c.ClusterGroups,
+		UserSecretsTargetClusters: c.UserSecretsTargetClusters,
+	}
+	var secrets []SecretConfig
+	for _, s := range c.Secrets {
+		var secret SecretConfig
+		if err := deepcopy.Copy(&secret, s); err != nil {
+			return nil, err
+		}
+		c.stripVaultPrefix(&secret)
+		secret.groupClusters()
+		secrets = append(secrets, secret)
+	}
+
+	target.Secrets = secrets
+	return json.Marshal(target)
+}
+
+func (s *SecretConfig) groupClusters() {
+	var secrets []SecretContext
+	for _, to := range s.To {
+		if len(to.ClusterGroups) > 0 {
+			sc := SecretContext{
+				ClusterGroups: to.ClusterGroups,
+				Namespace:     to.Namespace,
+				Name:          to.Name,
+				Type:          to.Type,
+			}
+			present := false
+			for _, context := range secrets {
+				if reflect.DeepEqual(context, sc) {
+					present = true
+					break
+				}
+			}
+			if !present {
+				secrets = append(secrets, sc)
+			}
+		} else {
+			// This cluster was not from a group, so nothing special needs to be done
+			secrets = append(secrets, to)
+		}
+	}
+
+	s.To = secrets
+}
+
+func (c *Config) stripVaultPrefix(s *SecretConfig) {
```

Comment:
The `c *Config` already contains many `*SecretConfig` - why are they getting passed in here? This data flow is unexpected.

Reply:
Yeah this is a bit of a weird one...I was more interested in making the function have a receiver to match all the others in the file than actually making it make any sense. Let me clean that up.